### PR TITLE
Implement PCA

### LINKED ISSUE #65: third option for specifying fitting intervals: fixed number of data points per interval
**I'm submitting a ...**
  - [ ] bug report
  - [x] feature request
  - [ ] code improvement request

## Expected behavior
**How should it work (with the bug fixed or the feature implemented)?**
When fitting data using the fitting module, it should be possible to specify the number of data points in a bin. 

## Actual behavior
**How does it currently work (with the bug causing problems or without the feature)?**
Right now only the number of intervals or the size of intervals can be specified.

### LINKED ISSUE #67: Implement the PCA model presented by Eckert-Gallup et al. (2016)
**I'm submitting a ...**
  - [ ] bug report
  - [x] feature request
  - [ ] code improvement request

## Expected behavior
**How should it work (with the bug fixed or the feature implemented)?**

Eckert-Gallup et al. (2016; DOI: 10.1016/j.oceaneng.2015.12.018) present a probabilistic model of sea states that can be used to construct environmental contours. It uses the the principal component analysis before fitting the model. This model shall be implemented.

An implementation is also available in WDRT's ESSC.py module: https://github.com/WEC-Sim/WDRT/blob/master/WDRT/ESSC.py#L897

### PR DESCRIPTION
**Pull request by kawindme**
Implements the contour creation according to Eckert-Gallup et al. (2016). (Fixes #67 )

To compare with the reference implementation (https://github.com/WEC-Sim/WDRT/blob/master/WDRT/ESSC.py) a new method to specify fitting intervals was implemented. (Fixes #65)

### CODE REVIEW

File: tests/test_pca.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+from viroconcom.utility import PCA
+from viroconcom.fitting import Fit
+from viroconcom.contours import IFormContour
+
+@pytest.mark.parametrize("buoy_num", [46022, 46023, 46025, 46050])
+def test_IFORM_contour_with_PCA_like_EckertGallupp2016(buoy_num):
+    # according to Eckert-Gallup2016: https://doi.org/10.1016/j.oceaneng.2015.12.018.
+    # and https://github.com/WEC-Sim/WDRT/blob/master/WDRT/ESSC.py
+    buoy_data = pd.read_csv(f"tests/testfiles/NDBC_buoy_{buoy_num}.csv")
```
[the remark this answers is not in the document]
Comment (kawindme) [reply tests/test_pca.py:19]: I think downloading the data during runtime would increase the test duration too much. 

When implementing this part some of the contours looked correct right away, while for others a big difference with the ESSC reference could be seen when plotting. Thus I would recommend to keep the data.

What advantage would it have to move the data to "datasets"?

File: viroconcom/distributions.py [hunk 2]
```diff
@@ -766,9 +768,69 @@ def _scipy_cdf(self, x, _, loc, scale):
...
     def _scipy_pdf(self, x, shape, loc, scale):
         return sts.norm.pdf(x, loc=loc, scale=scale)
+    
+    
+class InverseGaussianDistribution(ParametricDistribution):
+    """
+    An inverse Gaussian distribution.
+    
+    Let :math:`m` be the shape parameter and :math:`s` the scale parameter,
+    then the probability density function is defined as:
```
[the remark this answers is not in the document]
Comment (kawindme) [reply viroconcom/distributions.py:781]: The only reference would be scipy, as I derived that parameterization from the mentioned alternative to match scipy's shape, loc, scale scheme.

File: viroconcom/fitting.py [hunk 3]
```diff
@@ -954,9 +996,13 @@ def _get_fitting_values(sample, samples, name, dependency, index,
 
         # Define the data interval that is used for the fit.
         for i, step in enumerate(interval_centers):
-            mask = ((sorted_samples[:, 1] >= step - 0.5 * interval_width) &
-                    (sorted_samples[:, 1] < step + 0.5 * interval_width))
-            samples_in_interval = sorted_samples[mask, 0]
+            if samples_per_interval is not None:
+                # samples_in_interval = sample_chunks[i]
+                samples_in_interval = np.sort(sample_chunks[i][:, 0])
```
[the remark this answers is not in the document]
Comment (kawindme) [reply viroconcom/fitting.py:1001]: I am not entirely sure if the sorting is necessary, but as ESSC does it too, I assumed there was some purpose.

File: viroconcom/fitting.py [hunk 4]
```diff
@@ -580,16 +591,19 @@ def __init__(self, samples, dist_descriptions, timeout=None):
...
             Number of bins the data of this variable should be seperated for fits which depend
             upon it. If the number of bins is given, the width of the bins is determined
             automatically.
 
-        width_of_bins : float
+        width_of_intervals : float
             Width of the bins. When the width of the bins is given, the number of bins is
             determined automatically.
+            
+        samples_per_intervals : int
```
[the remark this answers is not in the document]
Comment (kawindme) [reply viroconcom/fitting.py:605]: points_per_interval is fine by me. I'll change it.

File: viroconcom/fitting.py [hunk 5]
```diff
@@ -954,9 +996,13 @@ def _get_fitting_values(sample, samples, name, dependency, index,
...
-            mask = ((sorted_samples[:, 1] >= step - 0.5 * interval_width) &
-                    (sorted_samples[:, 1] < step + 0.5 * interval_width))
-            samples_in_interval = sorted_samples[mask, 0]
+            if samples_per_interval is not None:
+                # samples_in_interval = sample_chunks[i]
+                samples_in_interval = np.sort(sample_chunks[i][:, 0])
+            else:
+                mask = ((sorted_samples[:, 1] >= step - 0.5 * interval_width) &
+                        (sorted_samples[:, 1] < step + 0.5 * interval_width))
+                samples_in_interval = sorted_samples[mask, 0]
```
[the remark this answers is not in the document]
Comment (kawindme) [reply viroconcom/fitting.py:1005]: It picks the points of the current interval from tthe sample. I did not change that part though, other than to add the else clause.

File: viroconcom/fitting.py [hunk 6]
```diff
@@ -1215,6 +1261,55 @@ def _get_distribution(self, dimension, samples, **kwargs):
...
+                                bounds = [(None, None), (None, None), (0, None)]
+                                res = minimize(error_func, p0, 
+                                               method="SLSQP", 
+                                               constraints=[ineq_cons],
+                                               bounds=bounds,
+                                               options={'ftol': 1e-9, "disp":False},
+                                               )
+                                param_popt = res.x
+                                    
+                                # if do_use_weights_for_dependence_function:
```
[the remark this answers is not in the document]
Comment (kawindme) [reply viroconcom/fitting.py:1287]: It was a reminder that do_use_weights_for_dependence_function is not implemented for poly2.
I'll delete a comment and add a NotImplementedError.

File: viroconcom/utility.py [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+        self.components = pca.components_
+        self.transformed = self._transform(data)
+        
+    def _transform(self, data):
+        #  sklearn pca.transform normalizes before transforming 
+        # X_transformed = np.dot(X, self.components_.T)
+        # here we do not normalize
+        Comp1_Comp2 = np.dot(data, self.components.T)
+        
+        # shift = abs(min(Comp1_Comp2[:, 1])) + 0.1 # ESSC style shift
```
[the remark this answers is not in the document]
Comment (kawindme) [reply viroconcom/utility.py:38]: This is a point were we divert from ESSC. ESSC always uses a shift, while we only use it if necessary. The commented line was to remind of the difference.

File: tests/test_pca.py [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+                          "dependency": (None, 0, 0),
+                          "functions": (None, "poly1", "poly2"),
+                          }
+
+    my_fit = Fit((Comp1_Comp2[:, 0], Comp1_Comp2[:, 1]),
+                 (dist_description_0, dist_description_1))
+
+    iform_contour = IFormContour(my_fit.mul_var_dist, Time_R, Time_SS, 1000)
+    
+    comp_coords = np.array(iform_contour.coordinates).T
```
[the remark this answers is not in the document]
Comment (kawindme) [reply tests/test_pca.py:40]: its the contour coordinates in pca space. I renamed it to pca_coords.

File: tests/test_pca.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ahaselsteiner) [reply tests/test_pca.py:20]: I would like to keep the repo light (in terms of MB and in terms of complexity). 
The "datasets" folder serves as the repository's place for environmental datasets. These datasets can also be used by users of viroconcom who download the full repository and who alter the existing example scripts. Currently, the datasets folder holds only 2 datasets that cover 1 year (to keep the repo light). Having four multi-year datasets are better datasets to work with. Thus, I think we should make them available in the intended place for datasets and where users can also easily find and use them for their own scripts.

File: viroconcom/fitting.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ahaselsteiner) [reply viroconcom/fitting.py]: Sounds good.

File: viroconcom/utility.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ahaselsteiner) [reply viroconcom/utility.py]: Ah - this is not clear to others who read the code. Could you comment on it in the code like in in your answer? Or alternatively, in the methods docstring at the top mention this difference.

**Pull request merged**